features/cellular/framework/AT/AT_CellularNetwork.cpp

```diff
@@ -429,11 +429,21 @@ nsapi_error_t AT_CellularNetwork::set_context_to_be_activated()
 
     // if user has defined user name and password we need to call CGAUTH before activating or modifying context
     if (_pwd && _uname) {
+#if defined(TARGET_WIO_3G)
```

Comment:
This is not a correct approach. Now I understand the context of your earlier question about  set_context_to_be_activated accessibility. I agree it does not make sense to copy paste activate_context method to your own class.
Please propose a solution where set_context_to_be_activated is changed to protected and virtual method, so you can override that in your class.

Reply:
What about introducing new protected virtual method for "user authentication" only and override that one?

Reply:
I can solve this issue with @AnttiKauppila approach soon, but @mirelachirica approach is too heavy for me. ( I don't want to change structures of shared/common code, since it could affect others platform. ) In addition, I don't want to support separating a function from AT command to response.

So, can I change set_context_to_be_activated() to protected and virtual method? @mirelachirica @jarvte

Reply:
I don't see any risk in changing the shared code something like this:

nsapi_error_t AT_CellularNetwork::set_context_to_be_activated()
{
    // try to find or create context with suitable stack
    if (!get_context()) {
        return NSAPI_ERROR_NO_CONNECTION;
    }

    // if user has defined user name and password we need to call CGAUTH before activating or modifying context
    nsapi_error_t ret_val = do_user_authentication();
    
    return ret_val;
}

nsapi_error_t AT_CellularNetwork::do_user_authentication()
{
    if (_pwd && _uname) {
        _at.cmd_start("AT+CGAUTH=");
        _at.write_int(_cid);
        _at.write_int(_authentication_type);
        _at.write_string(_uname);
        _at.write_string(_pwd);
        _at.cmd_stop();
        _at.resp_start();
        _at.resp_stop();
        if (_at.get_last_error() != NSAPI_ERROR_OK) {
            return NSAPI_ERROR_AUTH_FAILURE;
        }
    }
    return _at.get_last_error();
}

Reply:
I don't see any risk in changing the shared code something like this:

```
nsapi_error_t AT_CellularNetwork::set_context_to_be_activated()
{
    // try to find or create context with suitable stack
    if (!get_context()) {
        return NSAPI_ERROR_NO_CONNECTION;
    }

    // if user has defined user name and password we need to call CGAUTH before activating or modifying context
    nsapi_error_t ret_val = do_user_authentication();
    
    return ret_val;
}

nsapi_error_t AT_CellularNetwork::do_user_authentication()
{
    if (_pwd && _uname) {
        _at.cmd_start("AT+CGAUTH=");
        _at.write_int(_cid);
        _at.write_int(_authentication_type);
        _at.write_string(_uname);
        _at.write_string(_pwd);
        _at.cmd_stop();
        _at.resp_start();
        _at.resp_stop();
        if (_at.get_last_error() != NSAPI_ERROR_OK) {
            return NSAPI_ERROR_AUTH_FAILURE;
        }
    }
    return _at.get_last_error();
}
```
```

Reply:
But with this approach, what is the need of set_context_to_be_activated function?
Also calling inheritable function from private method is not a good design.
I think the best would be to rename set_context_to_be_activated() to do_user_authentication()
We can do this without breaking anything as it was a private method.
Also is get_context() check really needed in this function or should it be in a place where this function is called (activate_context)? Usually private functions are only called when the internal status is ok. And checking what activate_context does it makes perfect sense to move this check there.

Reply:
I agree with @AnttiKauppila proposals. One note still: not only rename set_context_to_be_activated() to do_user_authentication() but make it also protected virtual.

And the usage like:

```
nsapi_error_t AT_CellularNetwork::do_user_authentication()
{
    if (_pwd && _uname) {
        _at.cmd_start("AT+CGAUTH=");
        _at.write_int(_cid);
        _at.write_int(_authentication_type);
        _at.write_string(_uname);
        _at.write_string(_pwd);
        _at.cmd_stop();
        _at.resp_start();
        _at.resp_stop();
        if (_at.get_last_error() != NSAPI_ERROR_OK) {
            return NSAPI_ERROR_AUTH_FAILURE;
        }
    }

    return NSAPI_ERROR_OK;
}

nsapi_error_t AT_CellularNetwork::activate_context()
{
    _at.lock();

    nsapi_error_t err = NSAPI_ERROR_OK;

    // try to find or create context with suitable stack
    if(get_context()) {
        // try to authenticate user before activating or modifying context
        err = do_user_authentication();
    } else {
        err = NSAPI_ERROR_NO_CONNECTION;
    }

   if (err != NSAPI_ERROR_OK) {
.....
```

Reply:
@mirelachirica @AnttiKauppila I also needed to change accessibility of get_context(). Applied suggested changes.